session.go

```diff
@@ -65,6 +65,9 @@ type Session struct {
 	// or to send a header out directly.
 	sendCh chan sendReady
 
+	// sendDeadCh signals that the sending thread has completed.
+	sendDeadCh chan bool
```

Comment:
It looks like you only send `true` over this channel and the channel is never closed. This is not really considered a best practice and can be a bit confusing. A `chan bool` can carry a signal with three possible values (i.e true, false, or "closed" -- see https://play.golang.org/p/l7jzOU-kOFS).

The current code transmits a signal with 1 possible values (always true, never "closed"). It seems like the current code can send multiple signals. If repeated signals are truly desirable it may make sense to never close sendDeadCh. But you are still transmitting signal with one possible value using a data type with two possible values (disregarding closed channels for now). Other channels in this struct use `chan struct{}`. The `struct{}` type is a struct with no fields which takes up 0 bytes of memory and a variable of type `struct{}` has only one possible value (literal value `struct{}{}`). Because of this `chan struct{}` is viewed as a "pure" synchronization mechanism, it unblocks a receiver whenever a sender signals (close or send `struct{}{}`) and that's it. So if all you are trying to do is synchronize two goroutines the best practice is to use `chan struct{}`.

I've probably linked this talk multiple times but it's a good talk about concurrency in go and its relevant for this package: https://blog.golang.org/io2013-talk-concurrency